src/vm.rs

```diff
@@ -625,17 +565,17 @@ impl<'tx, 'transcript, 'gens> VM<'tx, 'transcript, 'gens> {
         Ok(Contract { predicate, payload })
     }
 
-    fn encode_output(&mut self, contract: Contract<'tx>) -> Vec<u8> {
+    fn encode_output(&mut self, contract: Contract) -> Vec<u8> {
         let mut output = Vec::with_capacity(contract.output_size());
-        encoding::write_point(&contract.predicate.0, &mut output);
+
+        encoding::write_point(&contract.predicate.to_point(), &mut output);
         encoding::write_u32(contract.payload.len() as u32, &mut output);
 
+        // can move all the write functions into type impl
         for item in contract.payload.iter() {
             match item {
                 PortableItem::Data(d) => {
-                    encoding::write_u8(DATA_TYPE, &mut output);
-                    encoding::write_u32(d.bytes.len() as u32, &mut output);
-                    encoding::write_bytes(d.bytes, &mut output);
+                    d.write(&mut output);
```

Comment:
Is there a reason for factoring out this code in data.write()? It seems to be pretty specific to contract serialization.

Reply:
not in particular — I thought it would make sense to have it as a method on the data type to read simpler, but it does seem like something that fits better just in the contract serialization function :+1: